File: frontend/src/lib/api/receiverStats.ts
/**
 * Typed client for the Receiver stats endpoints — `docs/API.md` §3.8
 * (roadmap slice 034): scorecard, time-series, range-by-bearing,
 * signal-distribution, lifetime.
 *
 * Reuses the `ApiV1Error`/`apiV1Fetch` pattern `lib/api/sightings.ts`
 * established for the external `/api/v1` surface's `{"error": {...}}`
 * envelope (§2.5) — duplicated rather than imported, the same call that
 * module itself makes relative to `aircraft.ts`.
 */
import { useQuery, type UseQueryResult } from "@tanstack/react-query";

/** `docs/API.md` §3.8's receiver-health summary — deliberately coarse; full
 * diagnostics is roadmap slice 042's scope. */
export type ReceiverHealth = "ok" | "no_stats" | "unknown" | "demo";

export interface ReceiverScorecard {
  current_visible: number;
  current_positioned: number;
  messages_per_sec: number | null;
  positions_per_sec: number | null;
  max_range_today_nm: number | null;
  max_range_ever_nm: number | null;
  unique_aircraft_today: number;
  unique_aircraft_since_t0: number;
  decoder_uptime_s: number | null;
  flightsite_uptime_s: number;
  health: ReceiverHealth;
}

/** SPEC §62's v1 chart catalog, minus the two endpoints with their own shape
 * (range-by-bearing, signal-distribution). */
export type ReceiverSeriesMetric =
  | "messages_per_sec"
  | "positions_per_sec"
  | "aircraft_count"
  | "max_range_nm"
  | "messages_total"
  | "positions_total"
  | "unique_aircraft";

export type ReceiverSeriesResolution = "high" | "hourly" | "daily";

export interface ReceiverSeriesPoint {
  t: string;
  value: number | null;
}

export interface ReceiverMetricSeries {
  metric: ReceiverSeriesMetric;
  resolution: ReceiverSeriesResolution;
  points: ReceiverSeriesPoint[];
}

export interface ReceiverBearingSector {
  /** Sector midpoint, degrees true. 0 is North, increasing clockwise. */
  bearing_deg: number;
  max_range_nm: number | null;
  at: string | null;
  icao: string | null;
}

export interface ReceiverRangeByBearing {
  sector_width_deg: number;
  today: ReceiverBearingSector[];
  ever: ReceiverBearingSector[];
}

export interface ReceiverSignalBucket {
  min_db: number;
  max_db: number;
  count: number;
}

export interface ReceiverSignalDistribution {
  from_ts: string | null;
  to_ts: string | null;
  bucket_width_db: number;
  buckets: ReceiverSignalBucket[];
  sample_count: number;
  min_db: number | null;
  max_db: number | null;
  avg_db: number | null;
}

export interface ReceiverMaxRangeRecord {
  nm: number;
  at: string;
  bearing_deg: number;
  icao: string | null;
}

export interface ReceiverBusiestDay {
  day: string;
  message_count: number;
}

export interface ReceiverFrequentAircraft {
  icao: string;
  registration: string | null;
  sighting_count: number;
}

export interface ReceiverCommonRecord {
  value: string;
  aircraft_count: number;
}

export interface ReceiverLifetimeStats {
  since: string | null;
  unique_aircraft: number;
  total_sightings: number;
  total_positions: number | null;
  total_messages: number | null;
  max_range: ReceiverMaxRangeRecord | null;
  peak_message_rate_per_sec: number | null;
  peak_position_rate_per_sec: number | null;
  max_simultaneous_aircraft: number | null;
  busiest_day: ReceiverBusiestDay | null;
  most_frequent_aircraft: ReceiverFrequentAircraft | null;
  common_type: ReceiverCommonRecord | null;
  common_model: ReceiverCommonRecord | null;
  common_operator: ReceiverCommonRecord | null;
}

interface ApiV1ErrorBody {
  error?: { code?: string; message?: string; detail?: unknown };
}

/** Thrown for any non-2xx response. `code` is the §2.5 machine-readable
 * slug, `null` when the response did not carry the documented envelope. */
export class ReceiverStatsApiError extends Error {
  readonly status: number;
  readonly code: string | null;

  constructor(status: number, body: ApiV1ErrorBody | undefined) {
    super(body?.error?.message ?? `Request failed with status ${status}`);
    this.name = "ReceiverStatsApiError";
    this.status = status;
    this.code = body?.error?.code ?? null;
  }
}

async function apiV1Fetch<T>(path: string): Promise<T> {
  const response = await fetch(path);
  if (!response.ok) {
    let body: ApiV1ErrorBody | undefined;
    try {
      body = (await response.json()) as ApiV1ErrorBody;
    } catch {
      body = undefined;
    }
    throw new ReceiverStatsApiError(response.status, body);
  }
  return (await response.json()) as T;
}

/** How often the scorecard is re-polled while the page is visible — short
 * enough that "current visible"/"messages per sec" reads as live without a
 * WebSocket subscription of its own. */
export const SCORECARD_POLL_MS = 5_000;

export function getReceiverScorecard(): Promise<ReceiverScorecard> {
  return apiV1Fetch<ReceiverScorecard>("/api/v1/receiver/scorecard");
}

export function useReceiverScorecardQuery(): UseQueryResult<ReceiverScorecard> {
  return useQuery({
    queryKey: ["receiver", "scorecard"],
    queryFn: getReceiverScorecard,
    refetchInterval: SCORECARD_POLL_MS,
  });
}

export interface ReceiverMetricSeriesParams {
  metric: ReceiverSeriesMetric;
  resolution: ReceiverSeriesResolution;
  /** Full ISO instant. Omit for the endpoint's resolution-sized default lookback. */
  from?: string;
  to?: string;
}

function seriesQuery(params: ReceiverMetricSeriesParams): string {
  const search = new URLSearchParams({
    metric: params.metric,
    resolution: params.resolution,
  });
  if (params.from !== undefined) {
    search.set("from", params.from);
  }
  if (params.to !== undefined) {
    search.set("to", params.to);
  }
  return search.toString();
}

export function getReceiverMetricSeries(
  params: ReceiverMetricSeriesParams,
): Promise<ReceiverMetricSeries> {
  return apiV1Fetch<ReceiverMetricSeries>(
    `/api/v1/receiver/metrics?${seriesQuery(params)}`,
  );
}

export function useReceiverMetricSeriesQuery(
  params: ReceiverMetricSeriesParams,
): UseQueryResult<ReceiverMetricSeries> {
  return useQuery({
    queryKey: ["receiver", "metrics", params],
    queryFn: () => getReceiverMetricSeries(params),
  });
}

export function getReceiverRangeByBearing(): Promise<ReceiverRangeByBearing> {
  return apiV1Fetch<ReceiverRangeByBearing>(
    "/api/v1/receiver/range-by-bearing",
  );
}

export function useReceiverRangeByBearingQuery(): UseQueryResult<ReceiverRangeByBearing> {
  return useQuery({
    queryKey: ["receiver", "range-by-bearing"],
    queryFn: getReceiverRangeByBearing,
  });
}

export interface ReceiverSignalDistributionParams {
  from?: string;
  to?: string;
  bucketWidthDb?: number;
}

function signalQuery(params: ReceiverSignalDistributionParams): string {
  const search = new URLSearchParams();
  if (params.from !== undefined) {
    search.set("from", params.from);
  }
  if (params.to !== undefined) {
    search.set("to", params.to);
  }
  if (params.bucketWidthDb !== undefined) {
    search.set("bucket_width_db", String(params.bucketWidthDb));
  }
  return search.toString();
}

export function getReceiverSignalDistribution(
  params: ReceiverSignalDistributionParams = {},
): Promise<ReceiverSignalDistribution> {
  const search = signalQuery(params);
  const path = search
    ? `/api/v1/receiver/signal-distribution?${search}`
    : "/api/v1/receiver/signal-distribution";
  return apiV1Fetch<ReceiverSignalDistribution>(path);
}

export function useReceiverSignalDistributionQuery(
  params: ReceiverSignalDistributionParams = {},
): UseQueryResult<ReceiverSignalDistribution> {
  return useQuery({
    queryKey: ["receiver", "signal-distribution", params],
    queryFn: () => getReceiverSignalDistribution(params),
  });
}

export function getReceiverLifetimeStats(): Promise<ReceiverLifetimeStats> {
  return apiV1Fetch<ReceiverLifetimeStats>("/api/v1/receiver/lifetime");
}

export function useReceiverLifetimeStatsQuery(): UseQueryResult<ReceiverLifetimeStats> {
  return useQuery({
    queryKey: ["receiver", "lifetime"],
    queryFn: getReceiverLifetimeStats,
  });
}

File: backend/src/flightsite/api/receiver_stats.py
"""Receiver stats API — query layer and pure helpers (roadmap slice 034).

``docs/API.md`` §3.8 publishes five endpoints: a scorecard, a generic
time-series, the range-by-bearing polar histogram, the signal-strength
distribution, and lifetime statistics. :mod:`flightsite.receiver_metrics` owns
the storage those first four partly read (SPEC §60/§64, ADR-0009); this module
owns everything that data does not answer on its own:

* **Unique aircraft** (today, since T0, and per-day for the ``unique_aircraft``
  chart) is deliberately *not* here — it is read straight from roadmap slice
  031's daily rollups via :class:`~flightsite.analytics.queries.AnalyticsQueries`
  (:meth:`~flightsite.api.context.LiveApiContext.receiver_scorecard` and
  :meth:`~flightsite.api.context.LiveApiContext.receiver_metric_series`), so
  this figure and the Analytics page's own "unique aircraft" stat tile answer
  from one query rather than two that could disagree.
* **Total sightings** and the lifetime "most/common" records below still are —
  the ``aircraft``/``sightings`` tables, which slice 031's rollups do not
  cover.
* **The signal-strength distribution** — per-sighting ``rssi_avg_db``
  (slice 052), explicitly *not* the raw-sample RSSI ``receiver_metrics``
  stores (see the note in ``flightsite.receiver_metrics``'s module docstring
  and roadmap slice 033's ``out_of_scope`` entry).
* **Lifetime "most/common" records** — joins over ``aircraft`` and resolved
  metadata that have nothing to do with metric retention.
* **The "ever" range-by-bearing reduction** — a pure fold of every stored
  daily record (``range_by_bearing_daily`` is retained indefinitely, §6.3)
  down to one all-time record per sector, using the same
  :func:`~flightsite.receiver_metrics.model.better_range` comparison
  production uses so the two can never disagree about what "further" means.

Every read here goes through :meth:`~flightsite.db.engine.Database.read_session`
(ADR-0001), exactly like :mod:`flightsite.api.history` and
:mod:`flightsite.api.sightings` — this module can never become a second
writer.
"""

from __future__ import annotations

import math
from collections.abc import Iterable, Mapping, Sequence
from dataclasses import dataclass
from datetime import date, timedelta
from typing import Any, Final

from sqlalchemy import ColumnElement, func, select

from flightsite.db import Database
from flightsite.db.models import Aircraft, AircraftMetadataResolved, Sighting
from flightsite.receiver_metrics.model import RangeRecord, better_range

#: Default RSSI histogram bucket width, dB. Coarse enough that a receiver's
#: typical signal spread fills a readable number of bars — SPEC §62 asks for a
#: distribution a person can read, not a per-tenth-of-a-dB table nobody acts on.
DEFAULT_SIGNAL_BUCKET_WIDTH_DB: Final = 3.0
MIN_SIGNAL_BUCKET_WIDTH_DB: Final = 0.5
MAX_SIGNAL_BUCKET_WIDTH_DB: Final = 20.0

#: SPEC §62's v1 chart catalog, minus the two endpoints with their own shape
#: (range-by-bearing, signal-distribution). ``unique_aircraft`` has no entry
#: here because it is never read from ``receiver_metrics_raw``/``*_summary``
#: at all — it comes from the analytics rollups (module docstring).
RAW_FIELD_FOR_METRIC: Final[Mapping[str, str]] = {
    "messages_per_sec": "messages_per_sec",
    "positions_per_sec": "positions_per_sec",
    "aircraft_count": "aircraft_visible",
    "max_range_nm": "max_range_nm",
}

#: The same catalog's summary-column counterpart, read once samples have been
#: folded into an hourly or daily bucket. ``aircraft_count`` reads the
#: bucket's *peak* simultaneous count (``aircraft_max``) rather than its
#: average: "simultaneous aircraft over time" (SPEC §62) is a chart of load,
#: and a peak is what a receiver operator actually wants to see per bucket.
SUMMARY_FIELD_FOR_METRIC: Final[Mapping[str, str]] = {
    "messages_per_sec": "msgs_per_sec_avg",
    "positions_per_sec": "pos_per_sec_avg",
    "aircraft_count": "aircraft_max",
    "max_range_nm": "max_range_nm",
    "messages_total": "messages_total",
    "positions_total": "positions_total",
}

#: Metrics with no raw-resolution representation: ``receiver_metrics_raw``
#: stores rates, not the totals these two name (see
#: :mod:`flightsite.receiver_metrics.aggregate`'s "Counts from rates" note).
SUMMARY_ONLY_METRICS: Final = frozenset({"messages_total", "positions_total"})

#: Default lookback applied when a series request omits ``from``, keyed by the
#: *effective* resolution — matching the frontend's 24h/7d/30d window selector
#: to the tier that can actually answer it: raw retains a couple of weeks at
#: most (ADR-0009), hourly and daily are permanent.
_MS_PER_DAY: Final = 24 * 3_600 * 1_000
DEFAULT_LOOKBACK_MS: Final[Mapping[str, int]] = {
    "high": 1 * _MS_PER_DAY,
    "hourly": 7 * _MS_PER_DAY,
    "daily": 30 * _MS_PER_DAY,
}


class ReceiverMetricQueryError(ValueError):
    """An unsupported ``metric``/``resolution`` pairing for the series endpoint.

    Raised for the two combinations ``docs/API.md`` §3.8 does not offer an
    answer to: ``unique_aircraft`` at anything but ``resolution=daily``, and a
    :data:`SUMMARY_ONLY_METRICS` metric at ``resolution=high``. The endpoint
    catches this and answers the §2.5 error envelope with a 400, the same
    pattern :func:`flightsite.airports.overlay.parse_bbox` uses for a
    malformed ``bbox``.
    """


def next_local_day(day: str) -> str:
    """The calendar day after ``day`` (``YYYY-MM-DD``), by date arithmetic alone.

    No timezone is involved: two consecutive local-day strings are always
    exactly one calendar day apart regardless of what a DST transition does to
    the *duration* between them, so this needs nothing
    :func:`~flightsite.receiver_metrics.aggregate.local_day` does not already
    guarantee about its own output.
    """
    return (date.fromisoformat(day) + timedelta(days=1)).isoformat()


def ever_ranges(rows: Iterable[tuple[str, RangeRecord]]) -> dict[int, RangeRecord]:
    """Reduce every stored ``(day, RangeRecord)`` pair to one record per sector.

    ``rows`` must be ordered oldest-day-first:
    :func:`~flightsite.receiver_metrics.model.better_range` keeps its
    ``current`` argument on a tie, so processing chronologically is what makes
    a tie resolve to *when the receiver first reached that far* — the same
    rule the lifetime max-range record uses.
    """
    best: dict[int, RangeRecord] = {}
    for _day, record in rows:
        bucket = record.bearing_bucket
        best[bucket] = better_range(best.get(bucket), record)
    return best


@dataclass(frozen=True, slots=True)
class SignalHistogramBucket:
    """One bar of the signal-strength distribution."""

    min_db: float
    max_db: float
    count: int


@dataclass(frozen=True, slots=True)
class SignalHistogram:
    """The signal-strength distribution — SPEC §62, from per-sighting RSSI."""

    bucket_width_db: float
    buckets: tuple[SignalHistogramBucket, ...]
    sample_count: int
    min_db: float | None
    max_db: float | None
    avg_db: float | None


def signal_histogram(
    values: Sequence[float], *, bucket_width_db: float = DEFAULT_SIGNAL_BUCKET_WIDTH_DB
) -> SignalHistogram:
    """Bucket ``values`` (per-sighting ``rssi_avg_db`` readings) into a fixed-width histogram.

    Pure and total: an empty ``values`` answers zero buckets and ``null``
    extremes rather than dividing by zero or fabricating a range (SPEC §39) —
    "no sighting in this window had a signal reading" is a real, ordinary
    first-run answer.
    """
    if bucket_width_db <= 0:
        raise ValueError("bucket_width_db must be greater than zero")
    if not values:
        return SignalHistogram(
            bucket_width_db=bucket_width_db,
            buckets=(),
            sample_count=0,
            min_db=None,
            max_db=None,
            avg_db=None,
        )

    low = min(values)
    high = max(values)
    start = math.floor(low / bucket_width_db) * bucket_width_db
    # At least one bucket even when every value is identical (high == low).
    bucket_count = max(1, math.ceil((high - start) / bucket_width_db))
    counts = [0] * bucket_count
    for value in values:
        index = min(max(int((value - start) / bucket_width_db), 0), bucket_count - 1)
        counts[index] += 1

    buckets = tuple(
        SignalHistogramBucket(
            min_db=start + index * bucket_width_db,
            max_db=start + (index + 1) * bucket_width_db,
            count=count,
        )
        for index, count in enumerate(counts)
    )
    return SignalHistogram(
        bucket_width_db=bucket_width_db,
        buckets=buckets,
        sample_count=len(values),
        min_db=low,
        max_db=high,
        avg_db=sum(values) / len(values),
    )


@dataclass(frozen=True, slots=True)
class MostFrequentAircraft:
    """The airframe with the most sightings — one entry of SPEC §63."""

    icao24: str
    registration: str | None
    sighting_count: int


@dataclass(frozen=True, slots=True)
class CommonRecord:
    """One "most common X" record — SPEC §63's type/model/operator entries."""

    value: str
    aircraft_count: int


class ReceiverStatsRepository:
    """Queries the scorecard, lifetime and signal-distribution endpoints need
    beyond what :class:`~flightsite.receiver_metrics.repository.MetricsRepository`
    already answers — see the module docstring.
    """

    __slots__ = ("_database",)

    def __init__(self, database: Database) -> None:
        self._database = database

    # -------------------------------------------------------------- charts

    async def total_sightings(self) -> int:
        """Every sighting ever recorded (SPEC §65: retained indefinitely)."""
        async with self._database.read_session() as session:
            return int(await session.scalar(select(func.count()).select_from(Sighting)) or 0)

    async def signal_values(self, *, from_ms: int | None, to_ms: int | None) -> tuple[float, ...]:
        """Per-sighting ``rssi_avg_db`` readings with ``started_at`` in ``[from_ms, to_ms]``.

        ``None`` bounds are unrestricted: an install with no explicit window
        gets the distribution over every sighting ever recorded, mirroring how
        ``GET /api/v1/sightings`` treats an absent ``from``/``to`` (§3.6).
        """
        conditions: list[ColumnElement[bool]] = [Sighting.rssi_avg_db.is_not(None)]
        if from_ms is not None:
            conditions.append(Sighting.started_ms >= from_ms)
        if to_ms is not None:
            conditions.append(Sighting.started_ms <= to_ms)
        statement = select(Sighting.rssi_avg_db).where(*conditions)
        async with self._database.read_session() as session:
            values = (await session.scalars(statement)).all()
        return tuple(float(value) for value in values if value is not None)

    # ------------------------------------------------------------- lifetime

    async def most_frequent_aircraft(self) -> MostFrequentAircraft | None:
        """The airframe with the most sightings, or ``None`` on an empty install."""
        statement = (
            select(
                Aircraft.icao24,
                Aircraft.sighting_count,
                AircraftMetadataResolved.registration,
            )
            .outerjoin(AircraftMetadataResolved, AircraftMetadataResolved.icao24 == Aircraft.icao24)
            .where(Aircraft.sighting_count > 0)
            .order_by(Aircraft.sighting_count.desc(), Aircraft.icao24.asc())
            .limit(1)
        )
        async with self._database.read_session() as session:
            row = (await session.execute(statement)).first()
        if row is None:
            return None
        return MostFrequentAircraft(
            icao24=row.icao24, registration=row.registration, sighting_count=row.sighting_count
        )

    async def common_type(self) -> CommonRecord | None:
        """The type code shared by the most distinct sighted airframes."""
        return await self._common(AircraftMetadataResolved.type_code)

    async def common_model(self) -> CommonRecord | None:
        """The model shared by the most distinct sighted airframes."""
        return await self._common(AircraftMetadataResolved.model)

    async def common_operator(self) -> CommonRecord | None:
        """The operator shared by the most distinct sighted airframes."""
        return await self._common(AircraftMetadataResolved.operator_name)

    async def _common(self, column: Any) -> CommonRecord | None:
        """The most frequent non-``null`` value of ``column``, among airframes
        this receiver has actually sighted.

        INNER JOINs to ``aircraft`` deliberately: ``aircraft_metadata_resolved``
        is a whole imported registry (FAA, Mictronics — SPEC §44's rarity
        checks read the same table), most of which this receiver has never
        heard. SPEC §63 asks for records about *this receiver's* lifetime, so
        the population here is airframes it has sighted, not the registry.
        """
        count = func.count(func.distinct(Aircraft.icao24))
        statement = (
            select(column.label("value"), count.label("aircraft_count"))
            .select_from(Aircraft)
            .join(AircraftMetadataResolved, AircraftMetadataResolved.icao24 == Aircraft.icao24)
            .where(column.is_not(None))
            .group_by(column)
            .order_by(count.desc(), column.asc())
            .limit(1)
        )
        async with self._database.read_session() as session:
            row = (await session.execute(statement)).first()
        if row is None:
            return None
        return CommonRecord(value=row.value, aircraft_count=row.aircraft_count)


__all__ = [
    "DEFAULT_LOOKBACK_MS",
    "DEFAULT_SIGNAL_BUCKET_WIDTH_DB",
    "MAX_SIGNAL_BUCKET_WIDTH_DB",
    "MIN_SIGNAL_BUCKET_WIDTH_DB",
    "RAW_FIELD_FOR_METRIC",
    "SUMMARY_FIELD_FOR_METRIC",
    "SUMMARY_ONLY_METRICS",
    "CommonRecord",
    "MostFrequentAircraft",
    "ReceiverMetricQueryError",
    "ReceiverStatsRepository",
    "SignalHistogram",
    "SignalHistogramBucket",
    "ever_ranges",
    "next_local_day",
    "signal_histogram",
]
